base/tps/src/main/java/org/dogtagpki/server/tps/rest/TPSCertService.java

```diff
@@ -162,6 +161,52 @@ public Response findCerts(String filter, String tokenID, Integer start, Integer
 
             return createOKResponse(response);
 
+        } catch (Exception e) {
+            throw new PKIException(e.getMessage());
+        }
+    }
+
+    private Response findAllCerts(String filter, Integer start, Integer size) {
+        String method = "TPSCertService:findAllCerts: ";
+        String msg = "";
+        TPSCertCollection response = new TPSCertCollection();
+
+        logger.debug("TPSCertService.findAllCerts({}, {}, {})", filter, start, size);
+
+        TPSEngine engine = TPSEngine.getInstance();
+        try {
+            TPSSubsystem subsystem = (TPSSubsystem) engine.getSubsystem(TPSSubsystem.ID);
+            TPSCertDatabase certDatabase = subsystem.getCertDatabase();
+            Iterator<TPSCertRecord> certRecs = certDatabase.findRecords(filter).iterator();
+            TokenDatabase tokenDatabase = subsystem.getTokenDatabase();
+            List<String> authorizedProfiles = getAuthorizedProfiles();
+            if (authorizedProfiles == null) {
+                msg = "authorizedProfiles null";
+                logger.debug("{}{}", method, msg);
+                throw new PKIException(method + msg);
+            }
+
+            int i = 0;
+            // skip to the start of the page
+            for (; i < start && certRecs.hasNext(); i++) certRecs.next();
+
+            if (certRecs.hasNext()) {
+                certRecs.forEachRemaining(certRecord -> {
+                    int j = 0;
+                    String tokenID = certRecord.getTokenID();
+                    TokenRecord tokenRecord = null;
+                    try {
+                        tokenRecord = tokenDatabase.getRecord(tokenID);
+                    } catch (Exception e) {
+                        // Proceed to next token if this one doesn't exist.
+                    }
+                    if (tokenRecord != null && authorizedProfiles.contains(UserResource.ALL_PROFILES) && j < size) {
+                        response.addEntry(createCertData(certRecord));
+                    }
+                });
+                response.setTotal(response.getEntries().size());
```

Comment:
Ideally the `total` should reflect the total number of entries that can be accessed by the user (the total number of entries matching the filter minus the ones that the user doesn't have permission to access). This is needed in order to page the results properly in the UI.

In line 159 it counts the total by going through the whole list. I'm not sure there's a more efficient way to do this though.

Reply:
OK, I did wonder what the point of the total was given that it is a `Collection` and finding the size is trivial. Thanks for the explanation!